core/src/main/java/hivemall/knn/similarity/DIMSUMMapperUDTF.java

```diff
@@ -0,0 +1,349 @@
+        if (colNorms == null || colProbs == null) {
+            final int numCols = colNormsOI.getMapSize(args[1]);
+
+            if (sqrtGamma == Double.POSITIVE_INFINITY) { // set default value to `gamma` based on `threshold`
+                if (threshold > 0.d) { // if `threshold` = 0, `gamma` is INFINITY i.e. always accept <j, k> pairs
+                    this.sqrtGamma = Math.sqrt(10 * Math.log(numCols) / threshold);
+                }
+            }
+
+            this.colNorms = new HashMap<Object, Double>(numCols);
+            this.colProbs = new HashMap<Object, Double>(numCols);
+            final Map<Object, Object> m = (Map<Object, Object>) colNormsOI.getMap(args[1]);
+            for (Map.Entry<Object, Object> e : m.entrySet()) {
+                Object j = e.getKey();
+                if (parseFeatureAsInt) {
+                    j = HiveUtils.asJavaInt(j);
+                } else {
+                    j = j.toString();
+                }
+
+                double norm = HiveUtils.asJavaDouble(e.getValue());
+                if (norm == 0.d) { // avoid zero-division
+                    norm = 1.d;
+                }
+
+                colNorms.put(j, norm);
+
+                double p = Math.min(1.d, sqrtGamma / norm);
+                colProbs.put(j, p);
+            }
+        }
+
+        if (parseFeatureAsInt) {
+            forwardAsIntFeature(row);
+        } else {
+            forwardAsStringFeature(row);
+        }
+    }
+
+    private void forwardAsIntFeature(@Nonnull Feature[] row) throws HiveException {
+        final int length = row.length;
+
+        Feature[] rowScaled = new Feature[length];
+        for (int i = 0; i < length; i++) {
+            int j = row[i].getFeatureIndex();
+
+            double norm = colNorms.get(j).doubleValue();
+            if (norm == 0.d) { // avoid zero-division
+                norm = 1.d;
+            }
+            double scaled = row[i].getValue() / Math.min(sqrtGamma, norm);
+
+            rowScaled[i] = new IntFeature(j, scaled);
+        }
+
+        final IntWritable jWritable = new IntWritable();
+        final IntWritable kWritable = new IntWritable();
+        final FloatWritable bWritable = new FloatWritable();
+
+        final Object[] forwardObjs = new Object[3];
+        forwardObjs[0] = jWritable;
+        forwardObjs[1] = kWritable;
+        forwardObjs[2] = bWritable;
+
+        for (int ij = 0; ij < length; ij++) {
+            int j = rowScaled[ij].getFeatureIndex();
+            double jVal = rowScaled[ij].getValue();
+
+            if (jVal != 0.d && rnd.nextDouble() < colProbs.get(j)) {
+                for (int ik = ij + 1; ik < length; ik++) {
+                    int k = rowScaled[ik].getFeatureIndex();
+                    double kVal = rowScaled[ik].getValue();
+
+                    if (kVal != 0.d && rnd.nextDouble() < colProbs.get(k)) {
+                        // compute b_jk
+                        bWritable.set((float) (jVal * kVal));
```

Comment:
Yeah, you are right. Even though `jVal` and `kVal` are "scaled" by `Math.min(sqrtGamma, norm)`, they could be very large in case that `gamma` is very small. Let's us double.